File: my-folder/0039-combination-sum/solution.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        backtrack(res, new ArrayList<>(), target, 0, candidates);
        return res;
    }

    private void backtrack(List<List<Integer>> res, List<Integer> temp, int target, int index, int[] nums) {
        if (target == 0) {
            res.add(new ArrayList<>(temp));
            return;
        }
        if (target < 0 || index >= nums.length) {
            return;
        }
        temp.add(nums[index]);
        backtrack(res, temp, target - nums[index], index, nums);
        temp.remove(temp.size() - 1);
        backtrack(res, temp, target, index+1, nums);
    }

}

File: my-folder/0039-combination-sum/solution.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def backtrack(candidates, target, temp, idx, res):
            if target == 0:
                res.append(temp[:])
                return
            if target < 0 or idx == len(candidates):
                return
            
            temp.append(candidates[idx])
            backtrack(candidates, target - candidates[idx], temp, idx, res)
            temp.pop()
            backtrack(candidates, target, temp, idx+1, res)
        
        res = []
        temp = []
        backtrack(candidates, target, temp, 0, res)
        return res
        
